### PR TITLE
JOB-89 Finished first iteration of slurp

### PR DESCRIPTION
**Pull request by dusktreader**
#### What
Adds a script to pull data from legacy jobbergate database into next-gen database.

#### Why
In order to move users from the legacy jobbergate apps into Armada, we need to be able to move the data seamlessly.

`Task`: https://app.clickup.com/t/18022949/JOB-89

---

#### Peer Review
Please follow the upstream omnivector documentation concerning
[peer-review guidelines](https://github.com/omnivector-solutions/Documentation/blob/main/Contributing/pr_review_standards.md#peer-review).

### CODE REVIEW

File: jobbergate-slurp/slurp/db_login.py [hunk 1]
```diff
@@ -0,0 +1,5 @@
+def login(is_legacy=typer.):
```
Comment (jamesbeedy) [inline jobbergate-slurp/slurp/db_login.py:1]: Is this a typo?

**Comment by jamesbeedy** [comment]
Really nice work here! Give me the weekend to look it over.

File: jobbergate-slurp/slurp/db_login.py (hunk 1 above)
Comment (dusktreader) [reply jobbergate-slurp/slurp/db_login.py:1]: yes, it sure is. Not sure how that got in there.

**Comment by matheushent** [comment]
does this code work in case the databases are in different VPCs?

**Comment by dusktreader** [comment]
> does this code work in case the databases are in different VPCs?

The code-base itself is agnostic of where the upstream databases are. You would have to set up a connection to an RDS instance in a VPC via an openvpn connection. If you needed to connect to a legacy database in one VPC and the nextgen database in another, you would have to configure openvpn client to connect to multiple vpns. I'm not sure how to do that, but this might be a start: https://forums.openvpn.net/viewtopic.php?t=22036

File: jobbergate-slurp/pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+py-buzz = "^2.1.3"
+
+[tool.poetry.dev-dependencies]
+
+[tool.poetry.scripts]
+slurp = "slurp.main:app"
+
+[build-system]
+requires = ["poetry-core>=1.0.0"]
+build-backend = "poetry.core.masonry.api"
```
Comment (jamesbeedy) [inline jobbergate-slurp/pyproject.toml:28]: +1

File: jobbergate-slurp/slurp/config.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    AUTH0_DOMAIN: str
+    AUTH0_AUDIENCE: str
+    AUTH0_CLIENT_ID: str
+    AUTH0_CLIENT_SECRET: str
+
+    class Config:
+        env_file = ".env"
+
+
+settings = Settings()
```
Comment (jamesbeedy) [inline jobbergate-slurp/slurp/config.py:28]: +1

File: jobbergate-slurp/slurp/connections.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+from slurp.config import settings
+
+
+def build_url(is_legacy=False):
+    env_prefix = 'NEXTGEN' if not is_legacy else 'LEGACY'
+    keys = ('user', 'pswd', 'host', 'port', 'name')
+    return "postgresql://{user}:{pswd}@{host}:{port}/{name}".format(
+        **{k: getattr(settings, f'{env_prefix}_DATABASE_{k.upper()}') for k in keys}
+    )
```
Comment (jamesbeedy) [inline jobbergate-slurp/slurp/connections.py:15]: ❤️

File: jobbergate-slurp/slurp/connections.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+            if not is_legacy:
+                logger.debug(f"Committed transaction for {database_url}")
+                connection.commit()
+            else:
+                logger.debug(f"Rolling back read-only transaction for {database_url}")
+                connection.rollback()
+        except Exception as err:
+             logger.error(f"Hit an error in transaction for {database_url}: {err}")
+             connection.rollback()
+             raise
```
Comment (jamesbeedy) [inline jobbergate-slurp/slurp/connections.py:38]: ❤️

File: jobbergate-slurp/slurp/migrators/applications.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+                owner_id=nextgen_interim_owner_id,
+                file=application["application_file"],
+                config=application["application_config"],
+                created=application["created_at"],
+                updated=application["updated_at"],
+            )
+        )
+        application_map[application["id"]] = nextgen_db.fetchone()["id"]
+    logger.debug("Finished migrating applications")
+    return application_map
```
Comment (jamesbeedy) [inline jobbergate-slurp/slurp/migrators/applications.py:51]: +1, great

File: jobbergate-slurp/slurp/migrators/job_scripts.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+                data=job_script["job_script_data_as_string"],
+                owner_id=nextgen_interim_owner_id,
+                application_id=nextgen_application_id,
+                created=job_script["created_at"],
+                updated=job_script["updated_at"],
+            )
+        )
+        job_scripts_map[job_script["id"]] = result.fetchone()["id"]
+    logger.debug("Finished migrating job_scripts")
+    return job_scripts_map
```
Comment (jamesbeedy) [inline jobbergate-slurp/slurp/migrators/job_scripts.py:49]: +1

File: jobbergate-slurp/slurp/migrators/job_submissions.py [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+                name=job_submission["job_submission_name"],
+                description=job_submission["job_submission_description"],
+                owner_id=nextgen_interim_owner_id,
+                job_script_id=nextgen_job_script_id,
+                slurm_job_id=slurm_job_id,
+                created=job_submission["created_at"],
+                updated=job_submission["updated_at"],
+            )
+        )
+    logger.debug("Finished migrating job_submissions")
```
Comment (jamesbeedy) [inline jobbergate-slurp/slurp/migrators/job_submissions.py:50]: +1

File: jobbergate-slurp/slurp/migrators/users.py [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+    logger.debug(f"Updated {result.rowcount} job_scripts")
+
+    logger.debug("Updating job_submissions with user map")
+    result = nextgen_db.execute("""
+        update job_submissions
+        set job_submission_owner_id = nextgen_user_id
+        from tmp_user_map
+        where job_submissions.job_submission_owner_id = tmp_user_map.legacy_user_id
+    """)
+    logger.debug(f"Updated {result.rowcount} job_submissions")
```
Comment (jamesbeedy) [inline jobbergate-slurp/slurp/migrators/users.py:74]: +1

File: jobbergate-slurp/slurp/pull_legacy.py [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+    result = legacy_db.execute("select * from job_scripts")
+    logger.debug(f"Fetched {result.rowcount} job_scripts")
+    return result.fetchall()
+
+
+def pull_job_submissions(legacy_db):
+    logger.debug("Fetching job_submissions from legacy database")
+    result = legacy_db.execute("select * from job_submissions")
+    logger.debug(f"Fetched {result.rowcount} job_submissions")
+    return result.fetchall()
```
Comment (jamesbeedy) [inline jobbergate-slurp/slurp/pull_legacy.py:33]: +1

**Review by jamesbeedy — APPROVED** [review]
I would like to see some docstrings, other then that I'm +1

**Comment by jamesbeedy** [comment]
What a useful tool this will be! Really nice work here!

**Comment by dusktreader** [comment]
> I would like to see some docstrings, other then that I'm +1

I will add some docstrings and then merge. Thanks for taking a look!

**Comment by matheushent** [comment]
This is really awesome @dusktreader. Nice work here!! In the future I think this can be a powerful open source tool as well

**Pull request merged**